autoprof/fit/lm.py

```diff
@@ -351,68 +463,113 @@ def take_low_rho_step(self):
+        h = torch.linalg.solve(
+            (
+                self.hess
+                + 1e-3
+                * self.L
+                * torch.eye(
+                    len(self.grad), dtype=AP_config.ap_dtype, device=AP_config.ap_device
+                )
+            )
+            * (
+                1
+                + self.L
+                * torch.eye(
+                    len(self.grad), dtype=AP_config.ap_dtype, device=AP_config.ap_device
+                )
+            )
+            ** 2
+            / (1 + self.L),
+            self.grad,
+        )
         return h
-    
-    def update_J_AD(self):
+
+    def update_J_AD(self) -> None:
         del self.J
         if "cpu" not in AP_config.ap_device:
             torch.cuda.empty_cache()
-        self.J = self.model.jacobian(torch.clone(self.current_state).detach(), as_representation = True, override_locked = False, flatten = True)
+        self.J = self.model.jacobian(
+            torch.clone(self.current_state).detach(),
+            as_representation=True,
+            override_locked=False,
+            flatten=True,
+        )
         if self.model.target.has_mask:
-            self.J[self.mask] = 0.
+            self.J[self.mask] = 0.0
         self.full_jac = True
-            
+
     @torch.no_grad()
-    def update_J_Broyden(self, h, Yp, Yph):
+    def update_J_Broyden(self, h, Yp, Yph) -> None:
         self.J = Broyden_step(self.J, h, Yp, Yph)
         if self.model.target.has_mask:
-            self.J[self.mask] = 0.
+            self.J[self.mask] = 0.0
         self.full_jac = False
 
     @torch.no_grad()
-    def update_hess(self):
+    def update_hess(self) -> None:
         if isinstance(self.W, float):
             self.hess = torch.matmul(self.J.T, self.J)
         else:
-            self.hess = torch.matmul(self.J.T, self.W.view(len(self.W),-1)*self.J)
-        self.hess += self.epsilon5 * torch.eye(len(self.current_state), dtype = AP_config.ap_dtype, device = AP_config.ap_device)
-            
+            self.hess = torch.matmul(self.J.T, self.W.view(len(self.W), -1) * self.J)
+        self.hess += self.epsilon5 * torch.eye(
+            len(self.current_state),
+            dtype=AP_config.ap_dtype,
+            device=AP_config.ap_device,
+        )
+
     @torch.no_grad()
-    def covariance_matrix(self):
+    def covariance_matrix(self) -> torch.Tensor:
         try:
             return torch.linalg.inv(self.hess)
         except:
-            AP_config.ap_logger.warning("WARNING: Hessian is singular, likely at least one model is non-physical. Will massage Hessian to continue but results should be inspected.")
-            self.hess += torch.eye(len(self.grad), dtype = AP_config.ap_dtype, device = AP_config.ap_device)*(torch.diag(self.hess) == 0)
+            AP_config.ap_logger.warning(
+                "WARNING: Hessian is singular, likely at least one model is non-physical. Will message Hessian to continue but results should be inspected."
```

Comment:
`message` should be `massage`

Reply:
I thought it was a typo
cuz massage is a different meaning right?

Reply:
Yes, massage is a different meaning. In this case I am using it to say that I modify the Hessian in a way that makes it not crash the code